File: dashboard/voice/transcription.py
"""Speech-to-text (Strategy).

WhisperCppTranscriber ports lettabot's working recipe: convert the uploaded blob
to 16 kHz mono PCM with ffmpeg, then run whisper-cli for a plain-text transcript.
The subprocess runner and the filesystem existence check are injectable so the
orchestration is unit-testable without real binaries.
"""
import os
import subprocess
import tempfile
from abc import ABC, abstractmethod
from pathlib import Path

from . import config


class TranscriptionError(Exception):
    pass


class TranscriptionStrategy(ABC):
    @abstractmethod
    def transcribe(self, audio_bytes: bytes, filename: str = "audio.webm") -> str:
        ...


def build_ffmpeg_args(ffmpeg_bin, src_path, wav_path):
    # -ar 16000 -ac 1 -c:a pcm_s16le  == what whisper.cpp expects.
    return [
        ffmpeg_bin, "-y", "-i", src_path,
        "-ar", "16000", "-ac", "1", "-c:a", "pcm_s16le", wav_path,
    ]


def build_whisper_args(binary_path, model_path, wav_path, output_base, language,
                       threads=None, prompt=None):
    args = [
        binary_path, "-m", model_path, "-f", wav_path,
        "-l", language, "-of", output_base, "-otxt", "-nt",
    ]
    if threads:
        args += ["-t", str(threads)]
    if prompt:
        # Initial prompt biases recognition toward our vocabulary (agent names).
        args += ["--prompt", prompt]
    return args


class WhisperCppTranscriber(TranscriptionStrategy):
    def __init__(self, binary_path, model_path, ffmpeg_path,
                 language="auto", threads=None, prompt=None,
                 runner=subprocess.run, exists=os.path.exists):
        self.binary_path = binary_path
        self.model_path = model_path
        self.ffmpeg_path = ffmpeg_path
        self.language = language
        self.threads = threads
        self.prompt = prompt
        self._run = runner
        self._exists = exists

    def transcribe(self, audio_bytes: bytes, filename: str = "audio.webm") -> str:
        if not audio_bytes:
            raise TranscriptionError("no audio data")
        for label, path in (("whisper binary", self.binary_path),
                             ("whisper model", self.model_path),
                             ("ffmpeg", self.ffmpeg_path)):
            if not self._exists(path):
                raise TranscriptionError(f"{label} not found: {path}")

        ext = (filename.rsplit(".", 1)[-1] or "webm").lower()
        with tempfile.TemporaryDirectory(prefix="dash-voice-") as tmp:
            src = os.path.join(tmp, f"source.{ext}")
            wav = os.path.join(tmp, "input.wav")
            out_base = os.path.join(tmp, "transcript")
            Path(src).write_bytes(audio_bytes)

            try:
                self._run(build_ffmpeg_args(self.ffmpeg_path, src, wav),
                          stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, timeout=120)
                self._run(build_whisper_args(self.binary_path, self.model_path, wav,
                                             out_base, self.language, self.threads, self.prompt),
                          stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, timeout=300)
            except subprocess.SubprocessError as exc:
                raise TranscriptionError(f"whisper.cpp failed: {exc}") from exc

            txt_path = out_base + ".txt"
            if not os.path.exists(txt_path):
                raise TranscriptionError("whisper.cpp produced no transcript")
            text = Path(txt_path).read_text(encoding="utf-8").strip()
            if not text:
                raise TranscriptionError("whisper.cpp returned an empty transcript")
            return text


def build_transcriber() -> TranscriptionStrategy:
    return WhisperCppTranscriber(
        config.WHISPER_CPP_BIN, config.WHISPER_MODEL_PATH, config.FFMPEG_BIN,
        language=config.WHISPER_LANGUAGE, threads=config.WHISPER_THREADS,
        prompt=config.WHISPER_PROMPT,
    )

File: dashboard/js/plans/voice-communication/specs/transcription.js
import { Status } from "../../interface-spec.js";

export const transcriptionSpec = {
  id: "transcription",
  name: "TranscriptionStrategy",
  group: "Shipped ports",
  tagline:
    "Audio bytes → raw text, via whisper.cpp. The most thoroughly tested port here.",
  status: Status.FINISHED,
  statusNote:
    "Complete and stable. 10 tests, plus 5 covering the pipeline that composes it.",
  responsibility: [
    "Convert an uploaded audio blob into raw text. One method, one job.",
    "The concrete implementation shells out: ffmpeg normalises whatever the browser recorded into 16 kHz mono PCM, then whisper-cli transcribes it. Both binaries are borrowed from lettabot rather than reinstalled, and every path is env-overridable.",
    "It is paired with, but separate from, CleanupStrategy. Transcription produces what was heard; cleanup fixes what was misheard. Keeping them apart is what lets voice_transcripts.json log both and make mis-heard agent names diagnosable.",
  ],
  contract: {
    language: "python",
    code: `TranscriptionStrategy(ABC)   voice/transcription.py

  transcribe(audio_bytes, filename) -> str      raises TranscriptionError

VoicePipeline   voice/pipeline.py    (composes transcribe -> cleanup)

  process(audio_bytes, filename) -> {raw_transcript, cleaned_text}

Pure arg-builders, tested independently of any subprocess:
  build_ffmpeg_args(...)     forces 16k mono PCM
  build_whisper_args(...)    text output, optional threads, optional prompt`,
    note: "WHISPER_PROMPT biases whisper toward the real agent names up front, so 'Mazda' is not heard as 'Melissa' in the first place.",
  },
  implementations: [
    {
      name: "WhisperCppTranscriber",
      kind: "current",
      file: "voice/transcription.py",
      note: "ffmpeg → 16k wav → whisper-cli (small.en since 2026-08-08).",
    },
    {
      name: "Pipecat WhisperSTTService",
      kind: "planned",
      file: "/home/adamsl/talking_agent_parts/",
      note: "The plan's Phase 2 would start here before writing any custom STT adapter.",
    },
  ],
  dependencies: {
    usedBy: ["VoicePipeline → POST /api/voice → MediaRecorderVoiceRecorder"],
    dependsOn: ["whisper-cli binary", "ffmpeg binary", "filesystem temp files"],
    note: "The arg-builders are pure functions, so the argument contract is tested without running either binary. Only two tests actually stub subprocess.",
  },
  developmentStatus: {
    done: [
      "Argument construction is pinned by tests: 16 kHz mono PCM, text output, optional thread count, optional initial prompt.",
      "A missing binary and an empty transcription both raise TranscriptionError rather than returning silence as success.",
      "The default model path is asserted against the model actually installed, so a model upgrade cannot silently drift from the test.",
      "VoicePipeline falls back to the raw transcript if cleanup fails — speech is never lost to a cleanup error.",
    ],
    gaps: [
      "Synchronous and blocking: a ~5s transcription occupies a request thread. Tolerable only because the HTTP server is threaded.",
      "No streaming/partial transcription, so push-to-talk cannot show text as you speak.",
      "No latency instrumentation — the plan's Phase 0 asks for measured end-of-speech→transcript timings, which have never been captured.",
    ],
  },
  tests: {
    files: [
      {
        path: "dashboard/tests/test_transcription.py",
        count: 10,
        proves:
          "Every argument the two binaries receive, prompt inclusion and omission, the happy path's ffmpeg-then-whisper ordering, and that a missing binary or empty output raises rather than returning ''.",
      },
      {
        path: "dashboard/tests/test_pipeline.py",
        count: 5,
        proves:
          "raw + cleaned are both returned, cleanup failure falls back to raw, and an empty upload is rejected before any work starts.",
      },
    ],
    untested: [
      "Behaviour on a corrupt or non-audio upload — ffmpeg's failure mode is not covered.",
      "Concurrency: two uploads transcribing at once share the temp-file naming scheme.",
    ],
    next: [
      "A corrupt-upload test asserting a clean TranscriptionError rather than a stack trace.",
      "Latency capture, so the Pipecat decision can be made against measured numbers instead of guesses.",
    ],
  },
  diagrams: [
    {
      title: "The push-to-talk leg",
      caption:
        "Two subprocesses behind one port method. Cleanup is a separate strategy so a cleanup failure cannot lose the transcript.",
      code: `sequenceDiagram
  participant VR as MediaRecorderVoiceRecorder
  participant API as POST /api/voice
  participant VP as VoicePipeline
  participant WT as WhisperCppTranscriber
  participant CU as CleanupStrategy

  VR->>API: audio blob
  API->>VP: process(bytes)
  VP->>WT: transcribe(bytes)
  WT->>WT: ffmpeg → 16k mono wav
  WT->>WT: whisper-cli → text
  WT-->>VP: raw transcript
  VP->>CU: clean(raw)
  alt cleanup succeeds
    CU-->>VP: cleaned text
  else cleanup fails
    VP-->>VP: fall back to raw
  end
  VP-->>VR: {raw_transcript, cleaned_text}`,
    },
  ],
  nextWork: [
    "Capture baseline latency (end-of-speech → transcript) on the live box.",
    "Add a corrupt-audio test.",
    "Leave the rest alone — this port is done and does not need work for its own sake.",
  ],
};
